src/main/java/com/springclass/library/services/impl/LibraryServiceImpl.java

```diff
@@ -0,0 +1,196 @@
+        bookDTO.setSubjects(subjectList);
+        return bookDTO;
+    }
+
+    @Override
+    public boolean deleteBook(Integer id) throws ServiceError {
+
+        if (id == null) {
+            throw new ServiceError(ErrorCode.INVALID_INPUT, "El ID del libro no puede ser nulo.");
+        }
+
+        if (!isNumeric(id.toString())) {
+            throw new ServiceError(ErrorCode.INVALID_INPUT, "El ID del libro debe ser un valor numérico.");
+        }
+
+        if(!bookRepository.existsById(id)){
+            throw new ServiceError(ErrorCode.RESOURCE_NOT_FOUND,"Libro no encontrado");
+        }
+        bookRepository.deleteById(id);
+        return true;
+    }
+
+    @Override
+    public BookDTO createBook(CreateBookDTO input) throws ServiceError {
+
+        if (input.getTitle() == null || input.getAuthor() == null || input.getSubjects() == null) {
+            throw new ServiceError(ErrorCode.INVALID_INPUT, "Los parámetros no pueden ser nulos.");
+        }
+
+        Author author = getOrCreateAuthor(input.getAuthor());
+        List<Subject> subjects = getOrCreateSubjects(input.getSubjects());
+
+        Book book = new Book();
+        book.setAutor(author);
+        book.setTitle(input.getTitle());
+        book.setSubjectList(subjects);
+
+        bookRepository.save(book);
+
+        return convertBookDto(book);
+    }
+
+    private Author getOrCreateAuthor(String authorName){
+        Optional<Author> optionalAuthor = authorRepository.findByName(authorName);
+        return optionalAuthor.orElseGet(() -> {
+            Author newAuthor = new Author();
+            newAuthor.setName(authorName);
+            return authorRepository.save(newAuthor);
+        });
+    }
+
+    private List<Subject> getOrCreateSubjects(List<String> subjects) {
+        List<Subject> subjectList = new ArrayList<>();
+
+        for(String subject : subjects){
+            Optional<Subject> optionalSubject = subjectRepository.findByTitle(subject);
+            if (optionalSubject.isPresent()) {
+                subjectList.add(optionalSubject.get());
+            } else {
+                Subject newSubject = new Subject();
+                newSubject.setTitle(subject);
+                subjectList.add(subjectRepository.save(newSubject));
+            }
+        }
+        return subjectList;
+    }
+
+    @Override
+    public List<BookDTO> findBooks(BookSearch search) throws ServiceError {
+
+        if (search == null || search.getTitle() == null && search.getAuthor() == null && search.getSubject() == null) {
+            List<Book> allBooks = bookRepository.findAll();
+            return allBooks.stream()
+                    .map(this::convertBookDto)
+                    .toList();
+        }
+
+        List<Book> bookList = bookRepository.findAll();
+        List<Book> filterBookList = bookList.stream()
+                .filter(book -> isBookMatchingCriteria(book, search))
+                .toList();
+        return filterBookList.stream()
```

Comment:
Aquí podrías haberlo ejecutado todo como una única llamada:
```suggestion
        return bookRepository.findAll()
              .filter(book -> isBookMatchingCriteria(book, search))
              .map(this::convertBookDto)
              .toList();
```